modules/markdown_utils.py

```diff
@@ -1,29 +1,128 @@
-from groq import Groq
-import os
-import streamlit as st
+def generate_refined_prompt(code_snippet, additional_instructions=""):
+    """
+    Generate a prompt for the model, with optional additional instructions injected.
+    """
+    base_prompt = f"""
+Translate the following Python model code into a very concise Markdown explanation of the underlying mathematical relationships and input uncertainties.
 
-def generate_prompt(code_snippet):
-    return f"""
-- Translate the following NumPy-based Python function into a markdown document that explains the following model defined in python:
 {code_snippet}
-- You must never re-print the model definition as python code.
-- You must keep your response to an abolute minimum - keep it extremely concise! If you are supplied with particularly large model with many inputs, feel free to obfuscate some of the maths when translating the model into latex/Markdown.
-- Present the mathematical equations using LaTex.
-- Tabulate (into ONE table!) the associated input uncertainties and their characteristics as detailed in the 'problem' dictionary.
 
-Provide the output in pure Markdown without additional explanations!
+**Requirements:**
+
+- Do not show or mention the Python code. Provide only a minimal, very concise explanation and the requested outputs.
+- Present all mathematical relationships as proper LaTeX equations using only inline ($...$) or display ($$...$$) math blocks.
+- If conditions are needed, use piecewise definitions with standard math inequalities, for example:
+  $$f(x)=\\begin{{cases}}
+  a & x \\geq b \\\\
+  c & x < b
+  \\end{{cases}}$$
+- Use only LaTeX inequalities (\\geq, \\leq) instead of code-like >= or <=.
+- All LaTeX must be syntactically correct: balanced braces, no extraneous characters, and directly renderable in standard Markdown.
+- Avoid complex LaTeX environments like align. Only use $...$ or $$...$$.
+- Provide exactly one Markdown table summarizing the input uncertainties, using 'problem'. Present variables in math mode, e.g. $Q$, and units (if any) as plain text outside math.
+- Output only Markdown:
+  1) One brief sentence introducing the model’s purpose,
+  2) The LaTeX equations,
+  3) The Markdown table of input symbols, their names (your best guess) and associated uncertainties.
+  
+No additional commentary or code excerpts. The output must be ready to render as-is in Markdown.
+
+Any deviation from these rules (e.g., using logical words in equations, showing code, extra braces) invalidates the response. Ensure correctness, simplicity, and complete compliance.
+
+{additional_instructions}
 """
+    return base_prompt
+
+def get_markdown_from_code(code_snippet, model_name='llama-3.1-8b-instant', max_attempts=10):
+    import os
+    from groq import Groq
+
+    forbidden_patterns = [
```

Comment:
This is the same list of forbidden patterns in `model_understanding`. I suggest to factor these settings in a single location.

Reply:
hi @mbaudin47 thanks for pointing these out: https://github.com/MLegkovskis/UncertaintyCat/pull/22

I have some sort of a auto-merge strategy setup on ci success, which has now been disabled.